consensus/ninja_tortoise.go

```diff
@@ -0,0 +1,456 @@
+			if blockCache[bChild].Layer() >= layer { //dont traverse too deep
+				if _, found := set[bChild]; !found {
+					set[bChild] = struct{}{}
+					stack.PushFront(bChild)
+				}
+			}
+		}
+	}
+	return layerCounter
+}
+
+func globalOpinion(v vec, layerSize uint32, delta float64) vec {
+	threshold := float64(GlobalThreshold*delta) * float64(layerSize)
+	if float64(v[0]) > threshold {
+		return Support
+	} else if float64(v[1]) > threshold {
+		return Against
+	} else {
+		return Abstain
+	}
+}
+
+func (ni *ninjaTortoise) updateCorrectionVectors(p votingPattern, bottomOfWindow mesh.LayerID) {
+	foo := func(x *mesh.Block) {
+		for _, bid := range ni.tEffectiveToBlocks[p] { //for all b who's effective vote is p
+			b := ni.blocks[bid]
+			if _, found := ni.tExplicit[b.Id][x.Layer()]; found { //if Texplicit[b][x]!=0 check correctness of x.layer and found
+				ni.Debug(" blocks pattern %d block %d layer %d", p, b.ID(), b.Layer())
+				if _, found := ni.tCorrect[b.Id]; !found {
+					ni.tCorrect[b.Id] = make(map[votingPattern]vec)
+				}
+				var vo vec
+				vo = ni.tVote[p][x.ID()]
+				ni.Debug("vote from pattern %d to block %d layer %d vote %d ", p, x.ID(), x.Layer(), vo)
+				ni.tCorrect[b.Id][p] = ni.tCorrect[b.Id][p].Add(vo.Negate()) //Tcorrect[b][x] = -Tvote[p][x]
+				ni.Debug("update correction vector for block %d layer %d , pattern %d vote %d for block %d ", b.ID(), b.Layer(), p, ni.tCorrect[b.Id][p], x.ID())
+			} else {
+				ni.Debug("block %d from layer %d dose'nt explicitly vote for layer %d", b.ID(), b.Layer(), x.Layer())
+			}
+		}
+	}
+
+	forBlockInView(ni.tPattern[p], ni.blocks, bottomOfWindow, foo)
+}
+
+func (ni *ninjaTortoise) updatePatternTally(newMinGood votingPattern, bootomOfWindow mesh.LayerID) {
+	ni.Debug("update tally pbase id:%d layer:%d p id:%d layer:%d", ni.pBase.id, ni.pBase.Layer(), newMinGood.id, newMinGood.Layer())
+	// bfs this sucker to get all blocks who's effective vote pattern is g and layer id i s.t pBase<i<p
+	//init p's tally to pBase tally
+	stack := list.New()
+	//include p
+	for _, b := range ni.tPattern[newMinGood] {
+		stack.PushBack(ni.blocks[b])
+	}
+
+	m := make(map[votingPattern]*correction)
+
+	foo := func(b *mesh.Block) {
+		if eff, found := ni.tEffective[b.ID()]; found {
+			if g, found := ni.tGood[eff.Layer()]; found && eff == g {
+				if _, f := m[g]; !f {
+					m[g] = &correction{}
+				}
+				m[g].add(ni.tCorrect[b.ID()][g])
+			}
+		}
+	}
+
+	forBlockInView(ni.tPattern[newMinGood], ni.blocks, ni.pBase.Layer()+1, foo) //ni.pBase.LayerID + 1 because we already copied pbase's votes
+
+	for idx := bootomOfWindow; idx < newMinGood.Layer(); idx++ {
+		g := ni.tGood[idx]
+		if corr, found := m[g]; found {
+			for b, v := range ni.tVote[g] {
+				tally := ni.tTally[newMinGood][b]
+				ni.Debug("correction vectors for g %d vector %d effective %d tally %d", g, corr.vector, corr.effectiveCount, tally)
+				tally = tally.Add(v.Multiplay(corr.effectiveCount).Add(corr.vector))
```

Comment:
This code also needs to be corrected once you have the corrected correction vector